xds/internal/xdsclient/authority_test.go

```diff
@@ -0,0 +1,364 @@
+	if !ok {
+		t.Fatalf("want a new controller to be build, got none")
+	}
+	// Want the server config for this authority.
+	wantConfig := serverConfigs[1]
+	if diff := cmp.Diff(ctrl.config, wantConfig, serverConfigCmpOptions); diff != "" {
+		t.Fatalf("controller is built with unexpected config, diff (-got +want): %v", diff)
+	}
+}
+
+// TestAuthorityShare covers that
+// - watch with the same authority name doesn't create new authority
+// - watch with different authority name but same authority config doesn't
+//   create new authority
+func (s) TestAuthorityShare(t *testing.T) {
+	overrideFedEnvVar(t)
+
+	ctrlCh, cleanupController := overrideNewController()
+	t.Cleanup(cleanupController)
+
+	client, err := newWithConfig(&bootstrap.Config{
+		XDSServer: serverConfigs[0],
+		Authorities: map[string]*bootstrap.Authority{
+			testAuthority:              {XDSServer: serverConfigs[1]},
+			testAuthority + "-another": {XDSServer: serverConfigs[1]}, // Another authority name, but with the same config.
+		},
+	}, defaultWatchExpiryTimeout, defaultIdleAuthorityDeleteTimeout)
+	if err != nil {
+		t.Fatalf("failed to create client: %v", err)
+	}
+	t.Cleanup(client.Close)
+
+	resourceName := buildResourceName(xdsresource.ClusterResource, testAuthority, testCDSName, nil)
+	ctrl1, ok1, _ := watchAndFetchNewController(t, client, resourceName, ctrlCh)
+	if !ok1 {
+		t.Fatalf("want a new controller to be build, got none")
+	}
+	// Want the server config for this authority.
+	wantConfig := serverConfigs[1]
+	if diff := cmp.Diff(ctrl1.config, wantConfig, serverConfigCmpOptions); diff != "" {
+		t.Fatalf("controller is built with unexpected config, diff (-got +want): %v", diff)
+	}
+
+	// Call the watch with the same authority name. This shouldn't create a new
+	// controller.
+	resourceNameSameAuthority := buildResourceName(xdsresource.ClusterResource, testAuthority, testCDSName+"1", nil)
+	ctrl2, ok2, _ := watchAndFetchNewController(t, client, resourceNameSameAuthority, ctrlCh)
+	if ok2 {
+		t.Fatalf("an unexpected controller is built with config: %v", ctrl2.config)
+	}
+
+	// Call the watch with a different authority name, but the same server
+	// config. This shouldn't create a new controller.
+	resourceNameSameConfig := buildResourceName(xdsresource.ClusterResource, testAuthority+"-another", testCDSName+"1", nil)
+	if ctrl, ok, _ := watchAndFetchNewController(t, client, resourceNameSameConfig, ctrlCh); ok {
+		t.Fatalf("an unexpected controller is built with config: %v", ctrl.config)
+	}
+}
+
+// TestAuthorityIdle covers that
+// - authorities are put in a timeout cache when the last watch is canceled
+// - idle authorities are not immediately closed. They will be closed after a
+//   timeout.
+func (s) TestAuthorityIdleTimeout(t *testing.T) {
+	overrideFedEnvVar(t)
+
+	ctrlCh, cleanupController := overrideNewController()
+	t.Cleanup(cleanupController)
+
+	const idleTimeout = 500 * time.Millisecond
```

Comment:
Can this be `100 * time.Millisecond` or lesser? With `500ms`, this test case alone will run for more than a second.